#1097 reduce memory overhead of the group by data structures

grandinj opened the pull request.

grandinj committed:
#1097 reduce memory overhead of the group by data structures

grandinj committed:
add nice explanation of why ValueHashMap is there

originally from Evgenij Ryazanov

grandinj committed:
accidentally removed some error checking

grandinj commented:
hmm, this actually makes it take longer on my machine

katzyn commented:
My explanation was not correct, Value is not a Comparable, but HashMap handles hashes in more complex way, if they have common less significant bits it's a big problem for ValueHashMap, but not a problem for HashMap.

But if all keys have exactly the same hash code ValueHashMap is still faster.

katzyn commented:
`HashMap` uses `hashCode ^ (hashCode >>> 16)` instead of raw `hashCode`, if less significant bits are still the same it builds a trees of entries with duplicate less significant bits but different whole hashes, or if hashes are the same too tries to use `Comparable.compareTo()` if possible (not our case) or other tricks. `HashMap` has a very complex logic that makes it more stable.

katzyn commented:
In the test case from the original issue reduced to 1,000,000 rows your branch with merged changes from the `master` is 10% faster on my system.

katzyn commented on h2/src/main/org/h2/util/ValueHashMap.java:
```diff
@@ -12,6 +12,21 @@
+ * entries, but slightly better behaviour with limited number of collisions,
+ * because collisions have no impact on non-colliding entires. HashMap in modern
+ * versions of Java also have the same overhead, but it builds a trees of keys
+ * with colliding hashes, that's why even if the all keys have exactly the same
```
`Value` is not a `Comparable` but trees that are based on the whole `hashCode` are still better.

katzyn commented on h2/src/main/org/h2/util/ValueHashMap.java:
```diff
@@ -12,6 +12,21 @@
+ * ValueHashMap is faster in typical cases, but may behave really bad in some
+ * cases. HashMap is slower in typical cases, but its performance does not
+ * degrade too much even in the worst possible case (if keys are comparable,
+ * like our Value class).
```
`Value` is not a `Comparable`, my mistake.

katzyn commented on h2/src/main/org/h2/command/dml/Select.java:
```diff
@@ -151,8 +163,47 @@ public void setGroupBy(ArrayList<Expression> group) {
+        if (index != null) {
+            if (currentGroupByExprData[index] != null) {
+                throw DbException.throwInternalError();
+            }
```
Looks like `assert` should be used instead. All existing callers of this method ensure that current value is `null` by theirself.

katzyn commented:
With test case from the original issue reduced to 2,000,000 rows retained memory for `ValueHashMap` is 900 MB before these changes and 459 MB with them.

On my system with 2,000,000 rows execution time is different on each run with both old and new implementation. New implementation was faster by a 3% when I compared results after a multiple runs, but deviation between runs is too high to be sure.

katzyn commented on h2/src/main/org/h2/command/dml/Select.java:
```diff
@@ -380,6 +431,10 @@ private void queryGroup(int columnCount, LocalResult result) {
             row = keepOnlyDistinct(row, columnCount);
             result.addRow(row);
         }
+        groupByData = null;
```
May be we need to clear `groupByData` in finally block? I'm not sure what happens in case of exception in the middle.

grandinj committed:
address review comments

grandinj committed:
cache lookups into exprToIndexInGroupByData

now the test is both faster and uses less memory than originally

grandinj committed:
Merge remote-tracking branch 'upstream/master' into 1097_slow_querygroup

katzyn commented on h2/src/main/org/h2/command/dml/Select.java:
```diff
@@ -151,8 +168,51 @@ public void setGroupBy(ArrayList<Expression> group) {
+     */
+    public void setCurrentGroupExprData(Expression expr, Object obj) {
+        if (cachedLookup == expr) {
+            assert currentGroupByExprData[cachedLookupIndex] != null;
```
I got an assertion error here in the original test case.

katzyn replied:
The logic is simple enough. Aggregate (or something) called `getCurrentGroupExprData()`, got a `null` value and sets a newly allocated `AggregateData` with `setCurrentGroupExprData()`. So `currentGroupByExprData[cachedLookupIndex]` can legally be `null` here.

katzyn replied:
Even more, it should be `null` here. Assertion should be reversed.

grandinj committed:
fix assertions

katzyn commented on h2/src/main/org/h2/command/dml/Select.java:
```diff
@@ -193,12 +193,12 @@ public Object getCurrentGroupExprData(Expression expr) {
         Integer index = exprToIndexInGroupByData.get(expr);
         if (index != null) {
-            assert currentGroupByExprData[index] != null;
+            assert currentGroupByExprData[index] == null;
```
Not here, this assertion was good.

grandinj replied:
I dont see how - if we setting a new value, there should be null in that entry, since nowhere do we ever overwrite a previous value

katzyn replied:
The value is already set a few lines above.

The execution path is not sane here.

katzyn replied:
If it was not set we can't enter this condition, probably it can be removed, but there is another big problem with empty map and non-null cached expression. I already mentioned it in another comment.

katzyn replied:
This method can be entered with the same expressions in cache and in parameter, but with empty map of all expressions.

Cached expression should be null if map of expressions is empty, but it's not sometimes.

grandinj committed:
add entries iterator to ValueHashMap

speeds up this query by about 30%

katzyn commented on h2/src/main/org/h2/command/dml/Select.java:
```diff
@@ -151,8 +168,51 @@ public void setGroupBy(ArrayList<Expression> group) {
+    public void setCurrentGroupExprData(Expression expr, Object obj) {
+        if (cachedLookup == expr) {
+            assert currentGroupByExprData[cachedLookupIndex] == null;
+            currentGroupByExprData[cachedLookupIndex] = obj;
```
We should return from the method here, but `exprToIndexInGroupByData` can be empty at this point.

If `exprToIndexInGroupByData` has such mapping we set `currentGroupByExprData[...]` again few lines below, that is not needed. But if `exprToIndexInGroupByData` is empty the mapping is inserted below.

Tests (with proper assertions) are fine, but logic seems to be weird enough. From my point of view we should not have empty map if cached expression and its index are set.

Also this optimization did not improve execution times on my tests.

May be it is not needed so much?

Also I think that this optimization can be done in another way. Some interface for allocation of `AggregateData` can be introduced. `Aggregate`, `JavaAggregate`, and `ExpressionColumn` can implement this interface. And instead of this setter method an another getOrSet method can lookup a value or allocate, set, and return a new value using this interface method from a caller. This way should be simpler.

katzyn replied:
I didn't notice before that this optimization sets value more than one time.

grandinj replied:
Yeah the variability in this test is a pain

katzyn commented on h2/src/main/org/h2/util/ValueHashMap.java:
```diff
@@ -174,6 +191,39 @@ public V get(Value key) {
+        }
+    }
+
+    private final class EntryIterator implements Iterator<Map.Entry<Value, V>> {
```
Java 7 does not have default methods, so `remove()` should be implemented too.

grandinj replied:
?? If we run it an older version of java, nobody will call remove() since it doesn't exist yet. If we run it on a newer version of Java, it will throw UnsupportedOperationException, which is fine (for our purposes)

katzyn replied:
Also `hasNext()` and `next()` methods violate the `Iterator` contract. The `next()` method should advance it to the next record, `hasNext()` can only inform whether the next invocation of `next()` will return a value or will throw a `NoSuchElementException`. It is legal to invoke `hasNext()` multiple times without invocation of `next()` and state of the iterator should not be changed.

But in this implementation `hasNext()` method used to traverse across map instead. That's not good.

katzyn replied:
> ?? If we run it an older version of java, nobody will call remove() since it doesn't exist yet.

This method is from JDK 1.2. We can't compile non-abstract iterators without `remove()` before Java 8. In Java 8 this method got default implementation.

grandinj committed:
missing return statement

grandinj committed:
forgot to clear the cachedLookup stuff

grandinj committed:
Iterator says we need to do the work in next()

and add remove() method for Java1.7 compat

grandinj committed:
switching to HashMap makes the performance a lot better

pretty close to the hsqldb performance, in fact, 20 seconds for H2 vs 17 seconds for HSQLDB

grandinj committed:
cachedlookup not worth it after all

was deceived by the variability of the test

grandinj committed:
walk this ValueHashMap using entries() method

since it seems to be more efficient than repeated lookups

katzyn commented:
`HashMap` makes some queries slower, but performance of `HashMap` is more stable. The only way to make it very slow is to craft keys with exactly the same hash code and without `Comparable` interface. I agree that here we can use `HashMap` for safety. May be we need to add `implements Comparable` to `Value` too, but this should be tested separately, comparison of many values is not so fast.

Performance with the test case from the original issue with 3,000,000 rows is good enough on my system too. Good work!

In H2 with these changes it takes 17 seconds in MVStore mode and 23 seconds in PageStore mode. HSQLDB executes this part in 15 seconds. With actual reading of values H2/MVStore takes 19-20 seconds and HSQLDB 16 seconds, this difference is expected.

katzyn commented on h2/src/main/org/h2/util/ValueHashMap.java:
```diff
@@ -174,6 +191,42 @@ public V get(Value key) {
+        }
+
+        @Override
+        public Map.Entry<Value, V> next() {
```
```Java
if (left <= 0) {
    throw new NoSuchElementException();
}
```

katzyn commented on h2/src/main/org/h2/util/ValueHashMap.java:
```diff
@@ -174,6 +191,42 @@ public V get(Value key) {
+                keysIndex++;
+                if (keys[keysIndex] != null && keys[keysIndex] != ValueNull.DELETED)
+                    return new DataUtils.MapEntry<Value, V>(keys[keysIndex], values[keysIndex]);
+            } while (true);
```
From my point of view `do … while (true)` is a very strange way to write an infinite loop. `for(;;)` is better. Or even `while(true) …` is better. It does not really matter, however.

Also we can read key from array only once per loop to reduce size of bytecode and improve readability:
```Java
                Value key = keys[++keysIndex];
                if (key != null && key != ValueNull.DELETED) {
                    return new DataUtils.MapEntry<>(key, values[keysIndex]);
                }
```

katzyn replied:
Class `DataUtils.MapEntry` can be removed and replaced everywhere with `java.util.AbstractMap.SimpleImmutableEntry` that is available since Java 6 / Android API 9.

katzyn commented on h2/src/main/org/h2/util/ValueHashMap.java:
```diff
@@ -174,6 +191,42 @@ public V get(Value key) {
+        return new EntryIterable();
+    }
+    
+    private final class EntryIterable implements Iterable<Map.Entry<Value, V>> {
```
An explicit no-arg constructor with default visibility is needed here to avoid synthetic access. It will replace automatic private constructor without any additional overhead.

katzyn commented on h2/src/main/org/h2/util/ValueHashMap.java:
```diff
@@ -174,6 +191,42 @@ public V get(Value key) {
+        }
+    }
+
+    private final class EntryIterator implements Iterator<Map.Entry<Value, V>> {
```
An explicit no-arg constructor with default visibility is needed here for the same reason.

katzyn reviewed (commented):
From my point of view changes in `Select` and related with them are fine.

But I have some complains and comments about `ValueHashMap`.

Visibility of `keys` and `values` fields of `ValueHashMap` should be set to default to avoid synthetic access in iterator and warnings from some compilers.

grandinj committed:
address review comments

avoid synthetic access methods

grandinj commented:
thanks for all the review comments

katzyn commented on h2/src/main/org/h2/util/ValueHashMap.java:
```diff
@@ -207,19 +211,25 @@ public V get(Value key) {
-            } while (true);
+                Value key = keys[keysIndex];
+                if (key != null && key != ValueNull.DELETED)
+                    return new AbstractMap.SimpleImmutableEntry<Value, V>(keys[keysIndex], values[keysIndex]);
```
It should be `(key, values[keysIndex])`.

The rest looks fine.

grandinj committed:
can use local var here
